BUG: sanity check `__array_interface__` number of dimensions

lvllvl opened the pull request:
Bug fix to #27949.

- Added a change to `numpy/_core/src/multiarray/ctors.c` to avoid segmentation fault
- added a test to `numpy/_core/tests/test_multiarray.py`. If the test passes there's no longer a segmentation fault




<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

lvllvl committed:
BUG: Prevent segmentation fault in np.asanyarray

seberg commented on numpy/_core/tests/test_multiarray.py:
```diff
@@ -10420,3 +10420,24 @@ def test_to_device(self):
         ):
             arr.to_device("cpu", stream=1)
+
+def test_asanyarray_excess_dimensions_raises():
```
```suggestion
def test_array_interface_excess_dimensions_raises():
```

seberg commented on numpy/_core/src/multiarray/ctors.c:
```diff
@@ -2250,6 +2250,12 @@ PyArray_FromInterface(PyObject *origin)
     else {
         n = PyTuple_GET_SIZE(attr);
+        if (n > NPY_MAXDIMS) {
+            PyErr_Format( PyExc_ValueError,
+                          "number of dimensions must be within [0, %d], got %d",
+                          NPY_MAXDIMS, n );
```
```suggestion
            PyErr_Format(PyExc_ValueError,
                         "number of dimensions must be within [0, %d], got %d",
                         NPY_MAXDIMS, n);
```
Just a nitpick but we never put spaces there.

There is actually another small bug (hard to test here).  `n` should be a `Py_ssize_t`.  I am honestly always surprised that we don't have compiler warnings for an implicit unsafe cast here...

seberg reviewed (commented):
Looks good, thanks.  If we touch this, I think we might as well also fix the bug for ridiculously large tuples here (don't care for a test, since that would require a lot of memory to run).

lvllvl commented:
> Looks good, thanks. If we touch this, I think we might as well also fix the bug for ridiculously large tuples here (don't care for a test, since that would require a lot of memory to run).

I wasn't able to find an open issue related to large tuples. Is this a known problem or should we open an issue for it?
Would you like this issue fixed within this PR or fixed in a separate PR?

seberg commented:
Universal test failures.  No there is no issue, just something to spot, `int` is the wrong integer type for `PyTuple_SIZE`, since it is basically the same code (and it is not like the diff is currently big), I would be happier to just include it in this PR, but works either way.

lvllvl commented:
> Universal test failures. No there is no issue, just something to spot, `int` is the wrong integer type for `PyTuple_SIZE`, since it is basically the same code (and it is not like the diff is currently big), I would be happier to just include it in this PR, but works either way.

@seberg ok I think I'll tackle it on a separate PR.
I think I fixed the test failures. `n` was already a variable in the function and it was of `type int`. So I changed my variable name and then reassigned it back to `n` after the fact so tests would build/pass.

seberg commented on numpy/_core/src/multiarray/ctors.c:
```diff
@@ -4158,6 +4171,12 @@ _array_fill_strides(npy_intp *strides, npy_intp const *dims, int nd, size_t item
+                    PyErr_SetString(PyExc_ValueError,
+                            "array is too big; overflow in _array_fill_strides.");
+                    return;
+                }
```
Why? Can we please not do this?  I am happy to check ridiculous tuple lengths one could claim them to be *technically* valid also since it is local to the `__array_interface__`.
But this just can't be valid.  Someone might as well pass a bad data pointer if they run into this.

lvllvl replied:
@seberg 
Would it be ok to focus on the original bug fix in this PR and then I'll address the large tuple issue separately? 

I wasn't able to reproduce the large tuple issue, but I can investigate it further in another PR. 

I implemented your other suggestions in regards to converting `n` and `i` into type `Py_ssize_t`.

seberg reviewed (commented):
First is a nit, but I don't like that new overflow check without reason much.
(I am not even sure it can be hit, but maybe it is possible with exact choice of itemsize and `prod(shape)`)

seberg reviewed (approved):
Thanks, @lvllvl.
